File: cs415-Parallel-Programming/cpp-java-consumer-producer/prodcons2.cpp
/*
    Brooke Czerwinski
    CS 415 Parallel Programming
    Assignment 1

    prodcons2.cpp

    C++ program to use parallel thread to add to and remove from a queue.
    Takes an int argument from command line and generates that many producer threads

    References used:
    https://en.cppreference.com/w/cpp/thread/condition_variable
    https://man7.org/linux/man-pages/man3/sched_getcpu.3.html
    https://www.cplusplus.com/reference/condition_variable/condition_variable/
    https://www.cplusplus.com/reference/condition_variable/condition_variable/wait/ 
*/
#include "queue.h"

#include <condition_variable>
#include <iostream>
#include <mutex>
#include <sched.h>
#include <thread>

using std::cerr;
using std::condition_variable;
using std::cout;
using std::endl;
using std::mutex;
using std::stoi;
using std::thread;
using std::unique_lock;


// Required Global Parameters
int     BUFSIZE = 20;       // queue capacity
int     NUMITEMS = 100;     // total number of data items

// Other Global Parameters
Queue               *aQueue;            // Queue pointers
int                 *consumerOutputArr;
mutex               mtx;                // Mutex variable
mutex               outputMtx;          // Mutex variable for consumer counts
condition_variable  cv;                 // Condition Variable
condition_variable  cv2;
bool                addToQueue = true;
bool                removeFromQueue = false;
bool                ioReady = false;
bool                outputArrReady = false;

// Function prototypes
void producer(int k);
void consumer(int k);                 
long checkInput(const int argc, const char input[]);     // Returns valid input or 0 if failed

int main(int argc, char *argv[])
{
    bool    errorFlag = false;
    long    numCons = 0;
    int     total = 0;
    
    // checkInput returns valid argument as an integer
    if (argc <= 2)
    {
        if (argc == 1)
        {
            numCons = 1;
        }
        else
        {
            numCons = checkInput(argc, argv[1]);
        }

        if (numCons == 0)
        {
            numCons = 1;
        }
        
        // Initialize queue
        aQueue = new Queue(BUFSIZE);

        // Initialize consumerOutputArr
        consumerOutputArr = new int[numCons];

        // Create producer thread
        thread  producerThread(producer, numCons);
        // Create consumer threads
        thread  consumerThreads[numCons];   // An array of consumer threads
        for (int i = 0; i < numCons; i++)
        {
            consumerThreads[i] = thread(consumer, i);
        }

        // Set the ioReady to true and notify one of the threads to start
        ioReady = true;
        cv.notify_one();
        outputArrReady = true;
        cv2.notify_one();

        // Join producer threads
        producerThread.join();
        // Join consumer threads
        for (int i = 0; i < numCons; i++)
        {
            consumerThreads[i].join();
        }

        // Final message
        cout << endl;
        cout << "Main: all done!" << endl;
        cout << "Consumer stats: [";

        for (int i = 0; i < numCons; i++)
        {
            cout << consumerOutputArr[i] << ",";
            total += consumerOutputArr[i];
        }
        cout << "] total = " << total << endl;
    
        // Release memory
        delete aQueue;
    }
    else
    {
        cerr << "Error, bad arguments. Please provide a single integer for the number of producer threads." << endl;
    }
    
    return 0;
}


// Checks to make sure program argument is a single valid integer
long checkInput(const int argc, const char input[])
{
    long numCons = 0;

    if (argc == 2)
    {
        numCons = strtol(input, NULL, 0);
    }

    return numCons;
}



void producer(int k)
{
    // Uses curly braces to establish a new scope so that "lck" will get 
    // destroyed (and lock released) when it goes out of scope.
    {
        // Obtain a lock on the i/o so messages don't get mixed together
        /*
            The lock is created using the mtx variable
            It will wait until getting a wakeup signal from the condition variable (cv)
        */
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Producer starting on core " << sched_getcpu() << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }
 
    for (int i = 1; i <= NUMITEMS; i++)
    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);
        // Wait until the condition variable sends a signal
        /*
            Notes for self:
            From https://www.cplusplus.com/reference/condition_variable/condition_variable/wait/

            If pred is specified (2), the function only blocks if pred returns false, 
            and notifications can only unblock the thread when it becomes true (which 
            is specially useful to check against spurious wake-up calls). 
            This version (2) behaves as if implemented as:
                    while (!pred()) wait(lck);
        */
        cv.wait(lck, []{return addToQueue;});

        cout << "Producer added " << i << " (qsz: " << aQueue->size() << ")" << endl;     
        // Add to the queue   
        aQueue->add(i);

        // If the queue has at least 1 item in it, set removeFromQueue to true
        if (aQueue->size() == 1)
        {
            removeFromQueue = true;
        }
        // If the queue is full, set addToQueue to false
        else if (aQueue->size() == BUFSIZE)
        {
            addToQueue = false;
        }
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }

    // Add k worth of -1's to the end of the queue
    for (int i = 0; i < k; i++)
    {
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return addToQueue;});
        aQueue->add(-1);

        // If the queue has at least 1 item in it, set removeFromQueue to true
        if (aQueue->size() == 1)
        {
            removeFromQueue = true;
        }
        // If the queue is full, set addToQueue to false
        else if (aQueue->size() == BUFSIZE)
        {
            addToQueue = false;
        }
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }

    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Producer ending" << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }
    
}



void consumer(int k)
{
    int totalRemoved = 0;
    int lastRemoved = 0;

    // Uses empty curly braces so that the unique lock will
    // die as it goes out of scope.
    // Lock the mutex so that the 
    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Consumer [" << k << "] starting on core " << sched_getcpu() << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }

    do
    {
        // Obtain a lock based on the mtx variable.
        unique_lock<std::mutex> lck(mtx);

        // wait for the condition variable to send a wakeup signal to continue
        // Secondary check: removeFromQueue must be true
        cv.wait(lck, []{return removeFromQueue;});

        // Output mesage and remove from queue at the same time
        lastRemoved = aQueue->remove();
        cout << "Consumer [" << k << "] rem'd " << lastRemoved << " (qsz: " << aQueue->size() << ")" << endl;

        // If queue is empty set removeFromQueue to false
        if (aQueue->size() == 0)
        {
            removeFromQueue = false;
        }
        // If queue has at least one remaining space in it, set addToQueue to true.
        else if (aQueue->size() == (BUFSIZE - 1))
        {
            addToQueue = true;
        }

        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();

        if (lastRemoved != -1)
        {
            totalRemoved++;
        }
    } while (lastRemoved != -1);

    {
        unique_lock<std::mutex> lck(outputMtx);
        cv2.wait(lck, []{return outputArrReady;});
        outputArrReady = false;
        consumerOutputArr[k] = totalRemoved;
        outputArrReady = true;
        cv2.notify_one();
    }

    {
        unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, []{return ioReady;});
        // Set the ioReady to false. Prevents competing i/o in case of spurrious wakeup
        ioReady = false;
        cout << "Consumer [" << k << "] ending" << endl;
        // Set the ioReady variable to true to set it up for the next waiting thread that gets to run
        ioReady = true;
        // Send a wakeup signal to one of the threads waiting using the condition variable (cv)
        cv.notify_one();
    }
}

File: cs415-Parallel-Programming/cpp-java-consumer-producer/ProdCons2.java
/*
    Brooke Czerwinski
    CS 415 Paralell Programming
    ProdCons2.java
    HW 1 

    Java-based producer/consumer program for parallel threads using a queue
    
    The program takes an optional command-line argument, numCons (number of consumers). If it is not
    provided, a default value of 1 is used.


    References used:
    https://stackoverflow.com/questions/5853167/runnable-with-a-parameter
    https://www.w3schools.com/java/java_lambda.asp
*/

import java.util.LinkedList;
import java.util.Queue;

public class ProdCons2{  

    // Required global parameters
    static int      BUFSIZE = 5;    // queue capacity
    static int      NUMITEMS = 100;  // total number of data items

    // Other global parameters
    static String   queueAccess = "queueAccess";    // Variable used for signalling/waiting
    static boolean  producerFinished = false;       // Whether or not the producer has finished adding to queue
    static boolean  queueFull = false;              // Whether or not the queue is full
    static boolean  queueEmpty = true;              // Whether or not the queue is empty
    static Queue<Integer> queue;                    // Queue to be used throughout the program
    static ConsumerClass[]  consumers;              // Holds the array of consumer threads
    static int numCons = 1;                         // Number of consumer threads. Can be changed with argument. Default is 1

    

    // ProducerClass which extends from thread
    // helps in the creation of producer threads.
    static class ProducerClass extends Thread
    {
        // Constructor
        ProducerClass() { super(); }

        // Override the run method
        @Override
        public void run() 
        {
            int count = 1;  // Number of items we have put into the queue (2-indexed)
            int num = 0;    // 

            System.out.println("Producer starting");

            // We are adding more than just NUMITEMS to the queue
            // After NUMITEMS have been added, we need to add numCons worth of -1's to the queue
            while (count <= (NUMITEMS + numCons))
            {
                synchronized(queueAccess)
                {
                    // If the queue is full the method waits on queueAccess to wake it
                    // Releases lock for now and release lock for now
                    while (queueFull == true)
                    {
                        try { queueAccess.wait(); }
                        catch (Exception e) { System.err.println(e.getMessage()); }
                    }

                    // Once the queue has room to add to it, we need to figure out what we're adding
                    // If we still have regular numbers to add we add our current count
                    if (count <= NUMITEMS)
                    {
                        num = count;
                    }
                    // Otherwise we add as many -1's to the queue as there are consumer threads
                    else
                    { 
                        num = -1; 
                    }

                    // Add the number to the qeueue, then print it out with the queue size.
                    queue.add(num);
                    System.out.println("Producer added " + num + " (qsz: " + queue.size() + ")");

                    // There is no way the queue is empty because we added something so set it to queueEmpty to false
                    queueEmpty = false;
                    // If the queue is now full after we've added something, set it to queueFull to true.
                    if (queue.size() >= BUFSIZE)
                    {
                        queueFull = true;
                    }

                    // Notify all threads to check to see if they can run again.
                    try { queueAccess.notifyAll(); }
                    catch (Exception e) { System.err.println(e.getMessage()); }
                }
                count++;
            }

            System.out.println("Producer ending");
        }
    }



    static class ConsumerClass extends Thread
    {
        int k;  // consumer thread id # given by main.

        ConsumerClass(int incoming) { 
            super(); 
            k = incoming;
        }

        @Override
        public void run()
        {
            int     last = 0;       // Last integer to be removed from the queue
            String  removed = "";   // String that will be displayed based on what was removed

            System.out.println("Consumer[" + k + "] running");

            // Keep removing things from the queue until a -1 has been removed.
            while (last != -1)
            {
                synchronized(queueAccess)
                {
                    while (queueEmpty == true)
                    {
                        try { queueAccess.wait(); }
                        catch (Exception e) { System.err.println(e.getMessage()); }
                    }

                    // Remove item from the queue and print out to the user
                    last = queue.remove();
                    if (last != -1)
                    {
                        removed = String.valueOf(last);
                    }
                    else
                    {
                        removed = "an ENDSIGN";
                    }
                    System.out.println("Consumer[" + k + "] rem'd " + removed + " (qsz: " + queue.size() + ")");

                    // Queue cannot be full because we just removed something from it, so set queueFull to false
                    queueFull = false;
                    // If the queue is now empty, set queueEmpty to true
                    if (queue.size() <= 0)
                    {
                        queueEmpty = true;
                    }

                    // Notify all threads waiting on queueAccess that they can try to get the lock again.
                    try { queueAccess.notifyAll(); }
                    catch (Exception e) { System.err.println(e.getMessage()); }
                }
            }

            System.out.println("Consumer[" + k + "] ending");
        }
    }

    public static void main(String args[])  
    {
        // If a parameter was provided, set numCons to that value
        if (args.length > 0)
        {
            numCons = Integer.parseInt(args[0]);  
        }

        // Initialize a queue
        queue = new LinkedList<>();
        
        // Create producer and consumer threads
        ProducerClass producer = new ProducerClass(); 
        consumers = new ConsumerClass[numCons]; 
        // Fill consumer thread array with ConsumerClass threads
        for (int i = 0; i < numCons; i++)
        {
            consumers[i] = new ConsumerClass(i);
        }

        // Start all producer and consumer threads
        producer.start();
        for (int i = 0; i < numCons; i++)
        {
            consumers[i].start();
        }

        // Join all producer and consumer threads back together
        try { producer.join(); }
        catch (Exception e) { System.err.println(e.getMessage()); }
        for (int i = 0; i < numCons; i++)
        {
            try { consumers[i].join(); }
            catch (Exception e) { System.err.println(e.getMessage()); }
        }    

        System.out.println("Main: all done!");
    }
}